tuist/tuist
# Prevent generation of redundant file elements

Resolves https://github.com/tuist/tuist/issues/514

### Short description 📝

File elements for all dependencies and products within a graph were always generated and included in projects without applying any filtering.
This led to the inclusion of redundant file elements in top level projects (for products & dependencies that they didn't require)

In this example:

Dependencies:
- Framework2 -> Framework3
- Framework4 -> Framework4
- Framework5 -> `ARKit.framework` however

The highlighted elements are redundant 

<img width="244" alt="Screenshot 2019-09-28 at 16 22 45" src="https://user-images.githubusercontent.com/11914919/65818741-790eaa80-e20c-11e9-9a61-b7af72e98689.png">


### Solution 📦

Leverage the business logic with the `Graph` to determine which file references are needed.


### Implementation 👩‍💻👨‍💻

- [x] Update `DependencyReference.product` to include product name
- [x] Add helper methods to `Graph` to allow listing all dependency references for a project
- [x] Update `ProjectFileElements` to leverage new helper
- [x] Update helpers to also resolve local swift packages

### Test Plan 🛠:

- Generate the fixture `fixtures/ios_app_with_frameworks` via `tuist generate`
- Open the generated Workspace
- Verify the `Framework3` project doesn't contain file references for `Framework5`
- Verify the `Framework3` project doesn't contain file references for `ARKit.framework`
- Verify the `Framework4` project doesn't contain file references for `ARKit.framework`
